diff --git a/cmake/Modules/ValkeySetup.cmake b/cmake/Modules/ValkeySetup.cmake
index bcce2bf1f..7807700be 100644
--- a/cmake/Modules/ValkeySetup.cmake
+++ b/cmake/Modules/ValkeySetup.cmake
@@ -238,6 +238,11 @@ endif ()
 if (APPLE)
     add_valkey_server_linker_option("-rdynamic")
     add_valkey_server_linker_option("-ldl")
+elseif(QNX)
+    add_valkey_server_linker_option("-rdynamic")
+    add_valkey_server_linker_option("-pthread")
+    add_valkey_server_linker_option("-lsocket")
+    add_valkey_server_linker_option("-lm")
 elseif (UNIX)
     add_valkey_server_linker_option("-rdynamic")
     add_valkey_server_linker_option("-pthread")
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index e4a81f6e4..eec3eb5de 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -17,7 +17,7 @@ add_dependencies(valkey-server generate_commands_def)
 add_dependencies(valkey-server generate_fmtargs_h)
 add_dependencies(valkey-server release_header)
 
-if (VALKEY_RELEASE_BUILD)
+if (VALKEY_RELEASE_BUILD AND NOT QNX)
     # Enable LTO for Release build
     set_property(TARGET valkey-server PROPERTY INTERPROCEDURAL_OPTIMIZATION TRUE)
 endif ()
diff --git a/src/Makefile b/src/Makefile
index 1ce928167..07a0aab57 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -227,6 +227,10 @@ ifeq ($(uname_S),Haiku)
 	FINAL_CFLAGS+= -DBSD_SOURCE
 	FINAL_LDFLAGS+= -lbsd -lnetwork
 	FINAL_LIBS+= -lpthread
+else
+ifeq ($(uname_S),QNX)
+	# QNX
+	FINAL_LIBS+= -lpthread -lsocket -lm 
 else
 	# All the other OSes (notably Linux)
 	FINAL_LDFLAGS+= -rdynamic
@@ -241,6 +245,7 @@ endif
 endif
 endif
 endif
+endif
 
 ifdef OPENSSL_PREFIX
 	OPENSSL_CFLAGS=-I$(OPENSSL_PREFIX)/include
diff --git a/src/fmacros.h b/src/fmacros.h
index c5c8ff4ad..d07461b17 100644
--- a/src/fmacros.h
+++ b/src/fmacros.h
@@ -48,7 +48,7 @@
 #define _ALL_SOURCE
 #endif
 
-#if defined(__linux__) || defined(__OpenBSD__)
+#if defined(__linux__) || defined(__OpenBSD__) || defined(__QNX__)
 #define _XOPEN_SOURCE 700
 /*
  * On NetBSD, _XOPEN_SOURCE undefines _NETBSD_SOURCE and
diff --git a/src/pqsort.c b/src/pqsort.c
index 31729cfbe..6a3de35cf 100644
--- a/src/pqsort.c
+++ b/src/pqsort.c
@@ -46,7 +46,9 @@ static inline char	*med3 (char *, char *, char *,
     int (*)(const void *, const void *));
 static inline void	 swapfunc (char *, char *, size_t, int);
 
+#ifndef min
 #define min(a, b)	(a) < (b) ? a : b
+#endif
 
 /*
  * Qsort routine from Bentley & McIlroy's "Engineering a Sort Function".
diff --git a/src/qnxfixes.h b/src/qnxfixes.h
new file mode 100644
index 000000000..1c42a7d3f
--- /dev/null
+++ b/src/qnxfixes.h
@@ -0,0 +1,8 @@
+#if defined(__QNX__)
+
+// linux has timeval in <time.h>, we have it in <sys/time.h>
+#include <sys/time.h> 
+
+#define IOV_MAX 1024 // taken from devs/sys/syslimits.h
+
+#endif
diff --git a/src/server.h b/src/server.h
index 25f01a31e..4c627fb96 100644
--- a/src/server.h
+++ b/src/server.h
@@ -33,6 +33,7 @@
 #include "fmacros.h"
 #include "config.h"
 #include "solarisfixes.h"
+#include "qnxfixes.h"
 #include "rio.h"
 #include "commands.h"
 #include "allocator_defrag.h"
@@ -1547,7 +1548,7 @@ typedef struct rdbSaveInfo {
 
 #define RDB_SAVE_INFO_INIT {-1, 0, "0000000000000000000000000000000000000000", -1}
 
-struct malloc_stats {
+struct _malloc_stats {
     size_t zmalloc_used;
     size_t process_rss;
     size_t allocator_allocated;
@@ -1820,7 +1821,7 @@ struct valkeyServer {
     long long stat_sync_partial_ok;                /* Number of accepted PSYNC requests. */
     long long stat_sync_partial_err;               /* Number of unaccepted PSYNC requests. */
     commandlog commandlog[COMMANDLOG_TYPE_NUM];    /* Logs of commands. */
-    struct malloc_stats cron_malloc_stats;         /* sampled in serverCron(). */
+    struct _malloc_stats cron_malloc_stats;         /* sampled in serverCron(). */
     long long stat_net_input_bytes;                /* Bytes read from network. */
     long long stat_net_output_bytes;               /* Bytes written to network. */
     long long stat_net_repl_input_bytes;           /* Bytes read during replication, added to stat_net_input_bytes in 'info'. */
diff --git a/src/util.c b/src/util.c
index 0631736c8..b09b31bf0 100644
--- a/src/util.c
+++ b/src/util.c
@@ -1225,7 +1225,7 @@ sds getAbsolutePath(char *filename) {
  * i.e timezone variables are linux specific.
  */
 long getTimeZone(void) {
-#if defined(__linux__) || defined(__sun)
+#if defined(__linux__) || defined(__sun) || defined(__QNX__)
     return timezone;
 #else
     struct timezone tz;
